ENH: add support for operator() in crackfortran.

dcaliste opened the pull request:
Some interface name may contains parenthesis when used
with operator, like:
 interface operator(==)
   module procedure my_type_equals
 end interface operator(==)

Make the end part properly detected, and store also
the operator ('==' in that case) in the name.

Also implement support to list the implemented by in
any interface declaration.

I would like to add some test for the new feature but I don't see how to write them in f2py/tests, some tips may be appreciated.

dcaliste committed:
ENH: add support for operator() in crackfortran.

Some interface name may contains parenthesis when used
with operator, like:
 interface operator(==)
   module procedure my_type_equals
 end interface operator(==)

Make the end part properly detected, and store also
the operator ('==' in that case) in the name.

Also implement support to list the implemented by in
any interface declaration.

dcaliste commented:
Hello, is there some interest in this ? Is it in a bad shape for integration ?

At least, it may be of some interest for [Sphinx Fortran](https://github.com/VACUMM/sphinx-fortran) for instance which is generating Fortran source code documentation using this crackfortran.py as a tokenizer. Currently, this operator interface is poorly documented.

mattip commented:
For tests, you can see gh-15035 which added tests for new parsing.

I am not a fortran user. Is the usage pattern you wish to support common? Should it be qualified with a version of fortran where support for this was added?

dcaliste commented:
Thanks for the link, I'll look at it during the week.

> I am not a fortran user. Is the usage pattern you wish to support common?

I think the operator overloading is part of the Fortran90 standard, which is now supported by all compilers (we have since Fortran95, 2003 and 2008 standards).

mattip commented:
Indeed, it seems operator overloading is supported in Fortran 90. Here is one [reference](http://www.pcc.qub.ac.uk/tec/courses/f90/stu-notes/F90_notesMIF_8.html#HEADING101), there are many available when searching for "spec fortran 90 operator overload". It seems assignment overloading is [also supported](http://www.pcc.qub.ac.uk/tec/courses/f90/stu-notes/F90_notesMIF_8.html#HEADING103) so this PR still needs:
- assignment overload parsing
- tests
- a release note in `doc/release/upcoming_changes`

If any of this is not clear, feel free to ask

dcaliste commented:
@mattip Thanks for looking on this PR. I'm on vacation leave at the moment, but I'll take care of adding the missing bits you mentioned after the 6th of January. Indeed, I forgot the assignment interface, my bad ;D I would have miss it when moving to some other Fortran sources I'm dealing with.

melissawm commented:
Hello @dcaliste  - are you interested in picking up work in this PR? Thanks!

dcaliste commented:
> assignment overload parsing
> tests
> a release note in doc/release/upcoming_changes

Thank you @mattip for the suggestions. I've added the assignment parsing, a test and a release note (and rebased on top of main).

> are you interested in picking up work in this PR?

Thank you for reminding me. I've updated the PR taking the suggestions of @mattip into account.

HaoZeke commented on doc/release/upcoming_changes/15006.new_feature.rst:
```diff
@@ -0,0 +1,7 @@
+crackfortran has support for operator and assignement overloading
+-----------------------------------------------------------------
+``crackfortran`` parser is now understanding operator and assignement
```
'is now understanding' --> 'now understands'

dcaliste replied:
Corrected.

HaoZeke commented on doc/release/upcoming_changes/15006.new_feature.rst:
```diff
@@ -0,0 +1,7 @@
+-----------------------------------------------------------------
+``crackfortran`` parser is now understanding operator and assignement
+defintions in a module. They are added in the ``body`` list of the
+module and contains a new key ``implementedby`` listing the names
```
'and contains' --> 'which contains'

dcaliste replied:
Changed.

HaoZeke reviewed (commented):
Minor grammar fixes aside; this looks good to me; however, since this does not correspond to an `f2py` feature yet; @dcaliste could you open an issue so this is highlighted? i.e. an issue which points out the lack of support in `f2py` for `operator()`. If you could fix the linter; I think this is ready

dcaliste commented:
Thank you @HaoZeke for your review. I've corrected the grammar mistakes and pleased the linter by breaking long lines.

About creating an issue about the lack of operator() support in `f2py`, I think, it's closely related to the lack of support for Fortran derived-types, as mentioned by @pearu in issue #14938. Indeed, there is no point to override operator and assignment for basic Fortran types. I may comment there the existence of this PR that may contribute to the effort for Fortran derived-type. Or do you prefer to open a new issue, specifically for this ?

dcaliste commented:
Thank you @HaoZeke, I've opened #19896 to follow the addition of operator() and assigment module procedures. I mentioned there that this PR is a step in that direction.

charris commented on doc/release/upcoming_changes/15006.new_feature.rst:
```diff
@@ -0,0 +1,7 @@
+crackfortran has support for operator and assignement overloading
+-----------------------------------------------------------------
```
"assignment", not "assignement"

dcaliste replied:
Oops, thanks for noticing.

HaoZeke commented on numpy/f2py/tests/test_crackfortran.py:
```diff
@@ -89,6 +89,41 @@ def test_defaultPublic(self, tmp_path):
+        f_path = tmp_path / "mod.f90"
+        with f_path.open('w') as ff:
+            ff.write(textwrap.dedent("""\
+            module foo
+              interface operator(.item.)
+                module procedure item_char, item_dict, item_dbl
+              end interface
+              interface operator(==)
+                module procedure dicts_are_equal
+              end interface
+              interface assignment(=)
+                module procedure get_value, get_integer, get_real
+              end interface
+            contains
+
+            end module foo
```
Could this be replaced with a full Fortran snippet? currently without a function or subroutine in contains this is incorrect.

dcaliste replied:
That's a good remark. I've updated the PR with a full Fortran snippet that compiles with GFortran and `-Wall -Werror`. Thanks for for pointing it out.

dcaliste commented:
I've corrected the test to include a valid Fortran snippet and also took the opportunity to rebase the PR.

HaoZeke commented on numpy/f2py/tests/test_crackfortran.py:
```diff
@@ -45,6 +45,77 @@ def test_defaultPublic(self, tmp_path):
+        f_path = tmp_path / "mod.f90"
+        with f_path.open('w') as ff:
+            ff.write(textwrap.dedent("""\
+            module foo
```
Very minor nit; we changed to a file based test interface in https://github.com/numpy/numpy/pull/20517.

dcaliste replied:
You're right. It's better with all this Fortran code in a file. Updated.

HaoZeke reviewed (commented):
Apart from the nit I think this is good to go in.

pearu reviewed (approved):
LGTM! Thanks, @dcaliste!

Btw, while reading this PR, I was unsure if crackfortran should be called a parser because it is not a parser in a conventional sense. For instance, crackfortran does not care if the input is syntactically correct or not while conventional parsers do. However, since it builds up an internal structure representing a part of Fortran code that is relevant only to wrapping Fortran to Python, I guess the parser may still be an appropriate term.. at least, I could not come up with a suggestion for a better term for crackfortran.

charris commented:
@HaoZeke Want to update your review?

HaoZeke reviewed (approved):
LGTM, thanks!

dcaliste commented:
> Btw, while reading this PR, I was unsure if crackfortran should be called a parser because it is not a parser in a conventional sense.

May I suggest "analyser" as a replacement ? I'm not a native English speaker, so it may not fit neither...

charris commented:
Thanks @dcaliste .

InessaPawson commented:
Hi-five on merging your first pull request to NumPy, @dcaliste! We hope you stick around! Your choices aren’t limited to programming – you can review pull requests, help us stay on top of new and old  issues, develop educational material, work on our website, add or improve graphic design, create marketing materials, translate website content, write grant proposals, and help with other fundraising initiatives. For more info, check out: https://numpy.org/contribute/. 
Also, consider joining our mailing list. This is a great way to connect with other cool people in our community and be part of important conversations that affect the development of NumPy: https://mail.python.org/mailman/listinfo/numpy-discussion
